### PR TITLE
Mirror privileges over data streams to their backing indices

### PR DESCRIPTION
**Pull request by danhermann**
This PR comprises the core of data streams security integration. It changes the authorization code to extend any privileges granted on a data stream to all of its backing indices. It also introduces an `includeDataStreams()` flag on any requests implementing `IndicesRequest` so that they can indicate how data streams should be considered during the authz process for each request. For requests where `includeDataStreams()` is `false`, authz will not include any data streams in the list of authorized indices for that request. For requests where `includeDataStreams()` is `true`, authz will include any matching data streams along with their respective backing indices in the list of authorized indices.

Note that @albertzaharovits's comment below (https://github.com/elastic/elasticsearch/pull/58381#issuecomment-649256434) includes another description of the way in which the authz code expands wildcards for requests that include data streams and those that do not.

Relates to #53100

### CODE REVIEW

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine update branch

**Review by albertzaharovits — COMMENTED** [review]
The core change looks good to me, but I'd like to sit on it a bit more. I'll get back to you tomorrow.

**Comment by albertzaharovits** [comment]
Me and @danhermann had a productive brainstorm session yesterday.

We worked on specifying the Security behaviour when the issued request (which might or might not "support data streams") contains wildcards (which Security must replace with concrete names from the authorised set, given the user's roles) and when the roles grant access to a data stream (or a data stream wildcard) compared to explicit backing indices. As a recap, we wish that when a role grants access to a namespace that covers an existing data stream it implicitly grants access to the backing indices of that data stream as well (but not the other way around, i.e. if the role grants access to some/all of the backing indices of a data stream it does NOT implicitly grant access to the data stream). This must work for all types of "index" privileges, i.e. indices:data* (read, index, delete, write, create, create_doc) and indices:admin* (view_index_metadata, manage, monitor,  maintenance). It's worth noting that this mirroring of privileges from the data stream to the backing index only happens if the data stream and the backing index exists in the first place, i.e. if the permission to create an index (or delete, or manage_follower/leader_index) is granted over some namespace it is NOT mirrored over the equivalent backing indices namespace (.ds-<namespace>).

We went over the following scenarios:

- the issued request contains wildcards that cover backing indices and the roles grant access to that namespace (or a subset of it) -> this is nothing new, existing authz covers this
- the issued request contains wildcards that cover data streams and the roles grant access to that namespace (or a subset of it). This forks into two cases:
  - the request supports data streams, in which case this is another of the ordinary case that is already covered by existing authz
  - the request does NOT support data streams, in which case wildcard expansion should not expand to concrete data stream names even if those are covered by the user's roles (this is because core expands willdcards for request that don't support data streams in a similar way)
- the issued request contains wildcards that cover data streams and the roles do NOT grant access to the covered data streams, but instead grant access to those data streams' backing indices. Permissions granted over indices do not mirror back to the data stream, so this case is once again a regular one, i.e. the requested namespace and the namespace covered by the role are treated as disjoint so the request is authorised for the empty name list.
- the issued request contains wildcards covering the backing indices of some data stream, but the role grants access only to the data stream. This is the important case. We must replace the wildcard in the request with the concrete names of the backing indices, provided that the backing indices represent a data stream that is covered by the role permission, and the request will be authorised as such. This is true for requests that do or do not support wildcards (@danhermann correct me if I'm wrong, I believe this diverges from what we discussed).  

----

We discussed implementation as well, but I was not able to get my head around all that enough to recommend an approach.

I now _believe_ we should change https://github.com/elastic/elasticsearch/blob/e33a0dfe77ad530db99bdd203434d16b23999be6/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java#L503 to include the backing index names of authorised data streams. That will work in conjunction with the change proposed in this PR, i.e. have the permission group check from https://github.com/elastic/elasticsearch/blob/e33a0dfe77ad530db99bdd203434d16b23999be6/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java#L216 deal with `indexAbstraction`s that although are not covered by the permission they are backing indices of data streams covered as such. Lastly, we should also modify [isIndexVisible](https://github.com/elastic/elasticsearch/blob/e33a0dfe77ad530db99bdd203434d16b23999be6/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java#L419) to hide concrete data streams when the request's wildcard covers one, but data streams are not supported for the particular request.

I only have one clarifying question: does `indexAbstraction.getIndices()` returns the backing indices for a data stream ? We need that to propagate permission from data stream to backing indices.

**Comment by martijnvg** [comment]
> does indexAbstraction.getIndices() returns the backing indices for a data stream ? 

Yes, if the index abstraction is of type data stream then it returns the corresponding backing indices.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@albertzaharovits, I've implemented the security integration as you suggested above and added unit tests. Also, I confirmed that along with the changes from https://github.com/elastic/elasticsearch/pull/57900, this will resolve https://github.com/elastic/elasticsearch/issues/57712.

I added a single override to `SearchRequest` to include data streams as that was necessary for some existing tests to pass, but I'll leave the updating of all the other appropriate request classes to another PR to keep the size of this one somewhat manageable.

If this matches what you wanted, I'll add some additional higher-level tests that validate all the new scenarios we discussed earlier.

I did notice one oddity in some manual testing that I was doing. If I create a user with a role limited to index read privileges on the pattern `b*`, I get the expected 404 if I attempt `GET b-foo/_search` for a non-existent `b-foo` index or data stream. However, the attempt `GET .ds-b-foo-000001/_search` returns a  `action [indices:data/read/search] is unauthorized for user [limited-user]` error and I haven't been able to track down the reason for that difference in behavior.

File: server/src/main/java/org/elasticsearch/action/IndicesRequest.java [hunk 1]
```diff
@@ -40,6 +40,13 @@
      */
     IndicesOptions indicesOptions();
 
+    /**
+     * Determines whether the action should be applied to data streams
+     */
```
Comment (albertzaharovits) [inline server/src/main/java/org/elasticsearch/action/IndicesRequest.java:45]: Suggestion: I would work on the naming and/or javadoc for this to make it clear that such requests over wildcards are meant to (not) expand to data stream names too.

**Comment by albertzaharovits** [comment]
@danhermann This is looking good! Glad to hear we're converging.

> I did notice one oddity in some manual testing that I was doing. If I create a user with a role limited to index read privileges on the pattern b*, I get the expected 404 if I attempt GET b-foo/_search for a non-existent b-foo index or data stream. However, the attempt GET .ds-b-foo-000001/_search returns a  action [indices:data/read/search] is unauthorized for user [limited-user] error rather than the expected 404 and I haven't been able to track down the reason for that difference in behavior.

This is expected behaviour not introduced by this PR.
GET .ds-b-foo-000001/_search, like GET a-foo/_search, returns 403 unauthorized because the index, which does not exist, is not covered by the permission namespace of the user (if we would return 404 we would exposed information about unauthorized data). .ds-b-foo-000001 would be authorised only if it existed, because its parent data stream existed as well, which is covered by the permission namespace.
**Indices that are authorised only if they exist are sort of the "feature" of this change, although a bit strange, I agree.**

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@ruflin, this is the PR that includes the changes you need for ingest.

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by ruflin** [comment]
@neptunian @ph As soon as this gets merged, we need to clean up the `.ds-*` permissions in the Ingest Manager. Lets make sure we don't forget it.

**Comment by danhermann** [comment]
@albertzaharovits, let me know if you have any test suggestions here. I tried to include unit tests targeted narrowly at the behavioral changes in the three methods that were changed as well as top-level end-to-end tests. For the latter, I didn't see a good place to add those except for the REST tests.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by james-elastic** [comment]
👍

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java [hunk 2]
```diff
@@ -1551,25 +1554,54 @@ public void testDataStreamResolution() {
...
+    public void testDataStreamsAreNotVisibleWhenNotIncludedByRequest() {
+        final User user = new User("data-stream-tester2", "data_stream_test2");
+        GetAliasesRequest request = new GetAliasesRequest("*");
+        assertThat(request, instanceOf(IndicesRequest.Replaceable.class));
+        assertThat(request.includeDataStreams(), is(false));
+
+        final List<String> authorizedIndices = buildAuthorizedIndices(user, GetAliasesAction.NAME, request);
+        ResolvedIndices resolvedIndices = defaultIndicesResolver.resolveIndicesAndAliases(request, metadata, authorizedIndices);
+        assertThat(resolvedIndices.getLocal(), not(hasItem("logs-foo")));
+        assertThat(resolvedIndices.getLocal(), not(hasItem("logs-foobar")));
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:1579]: no backing indices as well.
But regular indices and aliases are visible. Same with backing indices.
(I would expand the namespace granted to by the role to ensure those entities are OK - i.e. I'm not suggesting another test, just a more complete one).

But I would also test a simple (not wildcard) data stream name.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java [hunk 3]
```diff
@@ -1551,25 +1554,54 @@ public void testDataStreamResolution() {
...
+    public void testDataStreamsAreVisibleWhenIncludedByRequest() {
+        final User user = new User("data-stream-tester2", "data_stream_test2");
+        SearchRequest request = new SearchRequest("*");
+        assertThat(request, instanceOf(IndicesRequest.Replaceable.class));
+        assertThat(request.includeDataStreams(), is(true));
+
+        final List<String> authorizedIndices = buildAuthorizedIndices(user, SearchAction.NAME, request);
+        ResolvedIndices resolvedIndices = defaultIndicesResolver.resolveIndicesAndAliases(request, metadata, authorizedIndices);
+        assertThat(resolvedIndices.getLocal(), hasItem("logs-foo"));
+        assertThat(resolvedIndices.getLocal(), hasItem("logs-foobar"));
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:1591]: Should include the backing indices as well, right?

**Review by albertzaharovits — APPROVED** [review]
LGTM, I think the tests you've added are minimal but good enough.

I've left a couple of test suggestions. I'm not insisting over them because the missing pieces are covered across separate test classes; the issue is we're really not methodical with tests here, and so the burden of structuring testing in this area should not befall on you.

**Comment by albertzaharovits** [comment]
Before merging, can you work on the description and the title as well?

Eg "Mirror privileges over data stream to their backing indices"

Also explain the wildcard behaviour for requests that don't work with wildcards.

**Comment by danhermann** [comment]
Thank you, @albertzaharovits. I'll make the changes you requested. I'm also happy to work on additional tests in a follow-up if there are others that you would like to have.

File: .idea/runConfigurations/Debug_Elasticsearch.xml [hunk 4]
```diff
@@ -1,11 +1,11 @@
 <component name="ProjectRunConfigurationManager">
   <configuration default="false" name="Debug Elasticsearch" type="Remote">
     <option name="USE_SOCKET_TRANSPORT" value="true" />
-    <option name="SERVER_MODE" value="true" />
+    <option name="SERVER_MODE" value="false" />
```
Comment (ywangd) [inline .idea/runConfigurations/Debug_Elasticsearch.xml:4]: @danhermann Is this change intended? Also the one for `AUTO_RESTART`?

Comment (danhermann) [reply .idea/runConfigurations/Debug_Elasticsearch.xml:4]: @ywangd, no I have a PR open to revert it here: https://github.com/elastic/elasticsearch/pull/58992

Comment (ywangd) [reply .idea/runConfigurations/Debug_Elasticsearch.xml:4]: Thanks
